File: lingshu/apps/api/app/services/power_system.py
from __future__ import annotations

import ast
import json
import re
from typing import Any

_FENCE = re.compile(r"^```(?:json|python|py)?\s*", re.I)
_LABEL = re.compile(r"^(?:修炼体系|战力体系|境界体系)\s*[：:]\s*")


def strip_power_text(raw: str) -> str:
    text = (raw or "").strip()
    text = _FENCE.sub("", text)
    text = re.sub(r"\s*```$", "", text)
    text = _LABEL.sub("", text)
    return text.strip()


def _extract_object(raw: str) -> str:
    start = raw.find("{")
    end = raw.rfind("}")
    if start >= 0 and end > start:
        return raw[start : end + 1]
    return raw


def _loads(text: str) -> Any | None:
    for candidate in (text, _extract_object(text)):
        if not candidate:
            continue
        try:
            return json.loads(candidate)
        except json.JSONDecodeError:
            pass
        try:
            return ast.literal_eval(candidate)
        except (ValueError, SyntaxError, MemoryError):
            pass
    return None


def _unwrap(data: Any, depth: int = 0) -> dict[str, Any] | None:
    if depth > 3:
        return None
    if isinstance(data, str):
        nested = _loads(strip_power_text(data))
        return _unwrap(nested, depth + 1) if nested is not None else None
    if not isinstance(data, dict):
        return None
    for key in ("power_system", "payload", "data", "result", "修炼体系", "战力体系"):
        inner = data.get(key)
        if inner is None:
            continue
        opened = _unwrap(inner, depth + 1)
        if opened and (opened.get("intro") or opened.get("stages") or opened.get("hard_rules")):
            return opened
    return data


def parse_power_system(raw: str) -> dict[str, Any] | None:
    text = strip_power_text(raw)
    if not text:
        return None
    blob = _unwrap(_loads(text))
    if not blob:
        return None
    intro = str(blob.get("intro") or blob.get("概述") or blob.get("总览") or "").strip()
    stages_raw = blob.get("stages") or blob.get("篇章") or []
    stages: list[dict[str, Any]] = []
    if isinstance(stages_raw, list):
        for item in stages_raw:
            if not isinstance(item, dict):
                continue
            realms = item.get("realms") or item.get("境界") or []
            if isinstance(realms, str):
                realms = [part.strip() for part in re.split(r"[\n；;]+", realms) if part.strip()]
            elif not isinstance(realms, list):
                realms = []
            stages.append(
                {
                    "stage": str(item.get("stage") or item.get("name") or item.get("title") or "").strip(),
                    "realms": [str(x).strip() for x in realms if str(x).strip()],
                    "level": str(item.get("level") or item.get("战力层级") or "").strip(),
                    "core": str(item.get("core") or item.get("核心") or "").strip(),
                    "promotion": str(item.get("promotion") or item.get("晋升") or "").strip(),
                    "notes": str(item.get("notes") or item.get("备注") or "").strip(),
                }
            )
    rules = blob.get("hard_rules") or blob.get("rules") or blob.get("硬规则") or []
    if isinstance(rules, str):
        rules = [part.strip() for part in re.split(r"[\n；;]+", rules) if part.strip()]
    elif not isinstance(rules, list):
        rules = []
    rules = [str(item).strip() for item in rules if str(item).strip()]
    if not intro and not stages and not rules:
        return None
    return {"intro": intro, "stages": stages, "hard_rules": rules}


def normalize_power_system(raw: str) -> str:
    parsed = parse_power_system(raw)
    if not parsed:
        return (raw or "").strip()
    return json.dumps(parsed, ensure_ascii=False, indent=2)


def coerce_power_payload(draft: dict[str, Any] | None) -> str:
    data = draft if isinstance(draft, dict) else {}
    value = data.get("power_system") or data.get("content")
    if isinstance(value, dict):
        return normalize_power_system(json.dumps(value, ensure_ascii=False))
    if isinstance(value, str) and value.strip():
        return normalize_power_system(value)
    if data.get("intro") is not None or data.get("stages") or data.get("hard_rules"):
        return normalize_power_system(
            json.dumps(
                {
                    "intro": data.get("intro") or "",
                    "stages": data.get("stages") or [],
                    "hard_rules": data.get("hard_rules") or [],
                },
                ensure_ascii=False,
            )
        )
    return ""


def format_power_system(raw: str, *, missing: str = "（未填）") -> str:
    text = (raw or "").strip()
    if not text:
        return missing
    parsed = parse_power_system(text)
    if not parsed:
        return text
    lines: list[str] = []
    if parsed.get("intro"):
        lines.append(str(parsed["intro"]))
    for stage in parsed.get("stages") or []:
        title = stage.get("stage") or "篇章"
        level = stage.get("level") or ""
        realms = "、".join(stage.get("realms") or [])
        head = f"【{title}】{level}".strip()
        if realms:
            head += f"｜境界：{realms}"
        lines.append(head)
        if stage.get("core"):
            lines.append(f"核心：{stage['core']}")
        if stage.get("promotion"):
            lines.append(str(stage["promotion"]))
        if stage.get("notes"):
            lines.append(f"备注：{stage['notes']}")
    rules = parsed.get("hard_rules") or []
    if rules:
        lines.append("硬规则：")
        lines.extend(f"- {rule}" for rule in rules)
    return "\n".join(line for line in lines if line).strip() or text


def list_realms(raw: str) -> list[str]:
    parsed = parse_power_system(raw)
    if not parsed:
        return []
    seen: set[str] = set()
    chain: list[str] = []

    def add(name: str) -> None:
        item = (name or "").strip()
        if not item or item in seen:
            return
        seen.add(item)
        chain.append(item)

    intro = str(parsed.get("intro") or "")
    match = re.search(r"「([^」]+)」", intro)
    if match:
        for part in re.split(r"[—–\-→＞>]", match.group(1)):
            add(part)
    for stage in parsed.get("stages") or []:
        for realm in stage.get("realms") or []:
            add(str(realm))
    return chain


def realm_lock_block(raw: str) -> str:
    realms = list_realms(raw)
    if not realms:
        return (
            "【修炼体系硬锁定】左边还没有完整境界表。写修炼时不要另起通用仙侠等级"
            "（禁止筑基、金丹、元婴、练气、先天、后天等）。"
        )
    names = "、".join(realms)
    return f"""【修炼体系硬锁定｜必须逐字遵守，优先级等同作者原话】
允许出现的境界/等级只有：{names}
- 禁止使用上表没有的任何修炼等级、战力称谓、境界别名（包括但不限于筑基、结丹、金丹、元婴、化神、练气、开光、辟谷、先天、后天、超凡、入圣）。
- 可以说「某某境前期/中期/后期/巅峰」，但大境界名称必须来自上表。
- 原文若写了表外等级，改写时必须改回上表对应境界，不得沿用错的。
- 旁白点评、配角自报、系统面板同样只能用上表，禁止自造一套。"""

File: lingshu/apps/web/src/lib/powerSystem.ts
export interface PowerStage {
  stage: string
  realms: string[]
  level: string
  core: string
  promotion: string
  notes: string
}

export interface PowerSystem {
  intro: string
  stages: PowerStage[]
  hard_rules: string[]
}

export interface LabeledLine {
  title: string
  body: string
}

function asText(value: unknown): string {
  return typeof value === 'string' ? value.trim() : value == null ? '' : String(value).trim()
}

function asList(value: unknown): string[] {
  if (Array.isArray(value)) return value.map((item) => asText(item)).filter(Boolean)
  const text = asText(value)
  if (!text) return []
  return text
    .split(/\n+|；|;/)
    .map((item) => item.replace(/^[-•]\s*/, '').trim())
    .filter(Boolean)
}

function stripFence(raw: string): string {
  let text = (raw || '').trim()
  text = text.replace(/^[“”"']+/, '').replace(/[“”"']+$/, '')
  text = text.replace(/^```(?:json|python|py)?\s*/i, '')
  text = text.replace(/\s*```$/, '')
  text = text.replace(/^(?:修炼体系|战力体系|境界体系)\s*[：:]\s*/, '')
  return text.trim()
}

function extractObject(raw: string): string {
  const start = raw.indexOf('{')
  const end = raw.lastIndexOf('}')
  if (start >= 0 && end > start) return raw.slice(start, end + 1)
  return raw
}

function pythonishToJson(raw: string): string {
  const src = raw
    .replace(/[‘’]/g, "'")
    .replace(/[“”]/g, '"')
    .replace(/＇/g, "'")
  let out = ''
  let i = 0
  while (i < src.length) {
    const c = src[i]
    if (c === "'" || c === '"') {
      const quote = c
      i += 1
      let chunk = ''
      while (i < src.length) {
        const ch = src[i]
        if (ch === '\\' && i + 1 < src.length) {
          const next = src[i + 1]
          if (next === '\n') {
            chunk += '\\n'
            i += 2
            continue
          }
          chunk += ch + next
          i += 2
          continue
        }
        if (ch === quote) {
          i += 1
          break
        }
        if (ch === '"') {
          chunk += '\\"'
          i += 1
          continue
        }
        if (ch === '\n') {
          chunk += '\\n'
          i += 1
          continue
        }
        if (ch === '\r') {
          chunk += '\\r'
          i += 1
          continue
        }
        if (ch === '\t') {
          chunk += '\\t'
          i += 1
          continue
        }
        chunk += ch
        i += 1
      }
      out += `"${chunk}"`
      continue
    }
    if (src.startsWith('None', i) && !/[A-Za-z0-9_]/.test(src[i + 4] || '')) {
      out += 'null'
      i += 4
      continue
    }
    if (src.startsWith('True', i) && !/[A-Za-z0-9_]/.test(src[i + 4] || '')) {
      out += 'true'
      i += 4
      continue
    }
    if (src.startsWith('False', i) && !/[A-Za-z0-9_]/.test(src[i + 5] || '')) {
      out += 'false'
      i += 5
      continue
    }
    out += c
    i += 1
  }
  return out.replace(/,\s*([}\]])/g, '$1')
}

function tryParseObject(raw: string): unknown {
  const candidates = [raw, pythonishToJson(raw)]
  for (const candidate of candidates) {
    try {
      return JSON.parse(candidate)
    } catch {
      /* next */
    }
  }
  return null
}

function unwrap(data: unknown, depth = 0): Record<string, unknown> | null {
  if (depth > 3) return null
  if (typeof data === 'string') {
    const nested = tryParseObject(extractObject(stripFence(data)))
    return nested ? unwrap(nested, depth + 1) : null
  }
  if (!data || typeof data !== 'object' || Array.isArray(data)) return null
  const blob = data as Record<string, unknown>
  for (const key of ['power_system', 'payload', 'data', 'result', '修炼体系', '战力体系']) {
    const inner = blob[key]
    if (inner == null) continue
    const opened = unwrap(inner, depth + 1)
    if (opened && (opened.intro || opened.stages || opened.hard_rules || opened.概述)) return opened
  }
  return blob
}

function toStages(raw: unknown): PowerStage[] {
  if (!Array.isArray(raw)) return []
  const stages: PowerStage[] = []
  for (const item of raw) {
    if (!item || typeof item !== 'object' || Array.isArray(item)) continue
    const row = item as Record<string, unknown>
    const stage = asText(row.stage || row.name || row.title)
    const realms = asList(row.realms || row.境界)
    if (!stage && !realms.length) continue
    stages.push({
      stage: stage || `第${stages.length + 1}篇`,
      realms,
      level: asText(row.level || row.战力层级),
      core: asText(row.core || row.核心),
      promotion: asText(row.promotion || row.晋升),
      notes: asText(row.notes || row.备注),
    })
  }
  return stages
}

export function parsePowerSystem(raw: string): PowerSystem | null {
  const text = stripFence(raw)
  if (!text) return null
  const loaded = tryParseObject(text) || tryParseObject(extractObject(text))
  const blob = unwrap(loaded)
  if (!blob) return null
  const stages = toStages(blob.stages || blob.篇章 || blob.realms)
  const intro = asText(blob.intro || blob.概述 || blob.总览)
  const rules = asList(blob.hard_rules || blob.rules || blob.硬规则)
  if (!intro && !stages.length && !rules.length) return null
  return { intro, stages, hard_rules: rules }
}

export function stringifyPowerSystem(data: PowerSystem): string {
  return JSON.stringify(
    {
      intro: data.intro,
      stages: data.stages,
      hard_rules: data.hard_rules,
    },
    null,
    2,
  )
}

export function extractRealmChain(data: PowerSystem): string[] {
  const fromIntro = data.intro.match(/「([^」]+)」/)
  if (fromIntro) {
    const parts = fromIntro[1]
      .split(/[—–\-→＞>]/)
      .map((item) => item.trim())
      .filter(Boolean)
    if (parts.length >= 3) return parts
  }
  const seen = new Set<string>()
  const chain: string[] = []
  for (const stage of data.stages) {
    for (const realm of stage.realms) {
      if (seen.has(realm)) continue
      seen.add(realm)
      chain.push(realm)
    }
  }
  return chain
}

export function splitLabeled(text: string, maxTitle = 16): LabeledLine {
  const match = text.trim().match(new RegExp(`^([^：:]{1,${maxTitle}})[：:]\\s*([\\s\\S]+)$`))
  if (match) return { title: match[1].trim(), body: match[2].trim() }
  return { title: '', body: text.trim() }
}

export function splitPromotion(text: string): LabeledLine[] {
  return text
    .split(/\n+/)
    .map((line) => line.trim())
    .filter(Boolean)
    .map((line) => splitLabeled(line, 12))
}

export function extractProtagonistStatus(data: PowerSystem): LabeledLine | null {
  for (const rule of data.hard_rules) {
    const row = splitLabeled(rule)
    if (/主角/.test(row.title) || /主角/.test(rule)) {
      return { title: row.title || '主角当前境界', body: row.body || rule }
    }
  }
  return null
}
